### PR TITLE
Use named arguments for JSON/YAML serialization

### LINKED ISSUE #6405: JSON::Serializable causes compilation error when combined with an initialize method with exactly 2 params
When you have a class that includes JSON::Serializable and also has an initialize method that takes in 2 params a compilation error is raised: 
```crystal
instance variable '@foo' of Bar must be String, not JSON::PullParser
```
See playgrounds below for examples
https://play.crystal-lang.org/#/r/4jka - works fine (notice 3 vars in init call)
https://play.crystal-lang.org/#/r/4jkb - compilation error (notices 2 vars in init call)
https://play.crystal-lang.org/#/r/4jkc - also works fine (notice 1 var in init call)

Seems to only be occurring when init takes 2 params, 1 and >2 work as expected.
Cheers

### PR DESCRIPTION
**Pull request by Sija**
Fixes #6405

### CODE REVIEW

File: src/yaml/mapping.cr [hunk 1]
```diff
@@ -120,7 +120,7 @@ module YAML
     #
     # FIXME: remove the dummy argument if we ever fix this.
 
-    def initialize(ctx : YAML::ParseContext, node : ::YAML::Nodes::Node, _dummy : Nil)
+    def initialize(ctx : YAML::ParseContext, node : ::YAML::Nodes::Node)
```
Comment (luislavena) [inline src/yaml/mapping.cr]: Shouldn't this use named arguments too?

**Comment by luislavena** [comment]
This might also need updates:

https://github.com/crystal-lang/crystal/blob/4f720ab5dbfc71aa82e2874aa7becc2e69d3d4ae/src/yaml/from_yaml.cr#L88

https://github.com/crystal-lang/crystal/blob/master/src/yaml/from_yaml.cr#L113

https://github.com/crystal-lang/crystal/blob/master/src/yaml/from_yaml.cr#L129

**Comment by Sija** [comment]
Hmmm, I dunno what to make out of this error:

```
Error in spec/std_spec.cr:2: while requiring "./std/**"

require "./std/**"
^

in spec/std/yaml/serializable_spec.cr:771: instantiating 'YAMLAttrModuleTest2:Class#from_yaml(String)'

    it { YAMLAttrModuleTest2.from_yaml(%({"phoo": 20, "bar": 30})).to_tuple.should eq({10, 20, 30}) }
                             ^~~~~~~~~

in src/yaml/from_yaml.cr:2: read before assignment to local variable ''

  new(ctx: YAML::ParseContext.new, node: parse_yaml(string_or_io))
  ^~~
```

**Comment by asterite** [comment]
No idea, just stick with code that compiles.

**Comment by kostya** [comment]
why for son you change 2 lines, but for yams 100500 lines.

this seems to work:
```diff
diff --git a/src/json/serialization.cr b/src/json/serialization.cr
index 46f0d3144..2b1050a1b 100644
--- a/src/json/serialization.cr
+++ b/src/json/serialization.cr
@@ -117,7 +117,7 @@ module JSON

       def self.new(pull : ::JSON::PullParser)
         instance = allocate
-        instance.initialize(pull, nil)
+        instance.initialize(pull: pull)
         GC.add_finalizer(instance) if instance.responds_to?(:finalize)
         instance
       end
@@ -132,7 +132,7 @@ module JSON
       end
     end

-    def initialize(pull : ::JSON::PullParser, dummy : Nil)
+    def initialize(*, pull : ::JSON::PullParser)
       {% begin %}
         {% properties = {} of Nil => Nil %}
         {% for ivar in @type.instance_vars %}
diff --git a/src/yaml/serialization.cr b/src/yaml/serialization.cr
index 1db40b557..c369e1441 100644
--- a/src/yaml/serialization.cr
+++ b/src/yaml/serialization.cr
@@ -123,7 +123,7 @@ module YAML

         ctx.record_anchor(node, instance)

-        instance.initialize(ctx, node, nil)
+        instance.initialize(ctx: ctx, node: node)
         GC.add_finalizer(instance) if instance.responds_to?(:finalize)
         instance
       end
@@ -138,7 +138,7 @@ module YAML
       end
     end

-    def initialize(ctx : YAML::ParseContext, node : ::YAML::Nodes::Node, dummy : Nil)
+    def initialize(*, ctx : YAML::ParseContext, node : ::YAML::Nodes::Node)
       {% begin %}
         {% properties = {} of Nil => Nil %}
         {% for ivar in @type.instance_vars %}
```

**Comment by Sija** [comment]
@kostya AFAIK CI complains but please go ahead and open a new PR to try it out.

**Pull request closed without merging**